### PR TITLE
fix(routing): correct redirects in dev

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

Closes https://github.com/withastro/astro/issues/10576
Closes PLT-2584

This PR brings **two** fixes.

1. The first fix was around the construction of `RouteData[]` during the scanning of the file system. When configure redirects, Astro requires to have the _old_ routes in the file system. The old routes are stored as `"page"` routes, but then we have the function that marks them as `"redirect"`. The end product was an array that contained both routes, and the first one was the `"page"`, which was picked first. The fix removes those routes if we have some redirects
2. The second fix provides some UX improvements to our users by returning an HTML response that emulates that one returned by a static build. This means that we return a `200` status code, which is a bit controversial, but it matches the redirect that happens in static HTML file.

## Testing

The current CI should pass. I manually tested the changes in a project.

### CODE REVIEW

File: packages/astro/src/core/render-context.ts [hunk 1]
```diff
@@ -155,6 +155,7 @@ export class RenderContext {
 			}
 			let response: Response;
 
+			console.log('route type', this.routeData.type);
```
Comment (florian-lefebvre) [inline packages/astro/src/core/render-context.ts:158]: Looks like you forgot it

File: packages/astro/src/vite-plugin-astro-server/route.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
 import type http from 'node:http';
 import {
 	DEFAULT_404_COMPONENT,
+	REDIRECT_STATUS_CODES,
```
Comment (ascorbic) [inline packages/astro/src/vite-plugin-astro-server/route.ts:4]: Looks like this isn't used

File: packages/astro/src/vite-plugin-astro-server/route.ts [hunk 3]
```diff
@@ -286,6 +288,30 @@ export async function handleRoute({
...
+			response.status < 400 &&
+			routeIsRedirect(route) &&
+			!config.build.redirects &&
+			pipeline.settings.buildOutput === 'static'
+		) {
+			// If we're here, it means that the calling static redirect that was configured by the user
+			// We try to replicate the same behaviour that we provide during a static build
+			response = new Response(
+				redirectTemplate({
+					delay: response.status === 302 ? 2 : 0,
```
Comment (ascorbic) [inline packages/astro/src/vite-plugin-astro-server/route.ts:302]: Is there something special about the 2 value?

Comment (ematipico) [reply packages/astro/src/vite-plugin-astro-server/route.ts]: I refactored the code to keep the computation of `delay` inside one place. This was already documented in the `generate.ts` file: https://developers.google.com/search/docs/crawling-indexing/301-redirects#metarefresh

Essentially, for static files, to simulate a `302` redirect - which are *temporary* -, we need an arbitrary delay so Google understand that they are real 302 status code pages.

**Review by florian-lefebvre — APPROVED** [review]

**Review by matthewp — APPROVED** [review]

File: packages/astro/src/vite-plugin-astro-server/route.ts (hunk 3 above)
Comment (ascorbic) [reply packages/astro/src/vite-plugin-astro-server/route.ts]: today I learned!

**Pull request merged**